lib/dataset/shape.cpp

```diff
@@ -128,30 +128,18 @@ DataArray concat(const scipp::span<const DataArray> das, const Dim dim) {
 }
 
 Dataset concat(const scipp::span<const Dataset> dss, const Dim dim) {
-  // Note that in the special case of a dataset without data items (only coords)
-  // concatenating a range slice with a non-range slice will fail due to the
-  // missing unaligned coord in the non-range slice. This is an extremely
-  // special case and cannot be handled without adding support for unaligned
-  // coords to dataset (which is not desirable for a variety of reasons). It is
-  // unlikely that this will cause trouble in practice. Users can just use a
-  // range slice of thickness 1.
   if (dss.empty())
     throw std::invalid_argument("Cannot concat empty list.");
   Dataset result;
-  if (dss.front().empty())
-    result.setCoords(Coords(concat(map(dss, get_sizes), dim),
-                            concat_maps(map(dss, get_coords), dim)));
   for (const auto &first : dss.front())
     if (std::all_of(dss.begin(), dss.end(),
                     [&first](auto &ds) { return ds.contains(first.name()); })) {
       auto das = map(dss, [&first](auto &&ds) { return ds[first.name()]; });
-      if (std::any_of(das.begin(), das.end(), [dim, &first](auto &da) {
-            return da.dims().contains(dim) || !equals_nan(da, first);
-          }))
-        result.setData(first.name(), concat(das, dim));
-      else
-        result.setData(first.name(), first);
+      result.setDataInit(first.name(), concat(das, dim));
     }
+  if (dss.front().empty())
+    return Dataset({}, Coords(concat(map(dss, get_sizes), dim),
+                              concat_maps(map(dss, get_coords), dim)));
   return result;
```

Comment:
Won't this return an invalid dataset if the `std::all_of` above is always `false`?

Reply:
Why? It would still merge and return the coords.

Reply:
But if there are no items to merge?

Reply:
Do you mean when the inputs don't have any overlapping data items and none of them have coords? Then yes, it would return an invalid dataset and that is bad.

Reply:
Correct me if I am wrong, but even if there *are* coords (but no overlapping data items) you will get invalid?

Reply:
Why? Doesn't this code merge all coords of all inputs?

Reply:
Which code?

Reply:
The one you highlighted. It calls gets all coords of all datasets and merges them. This seems independent of whether the `dss` contain data or not.

Reply:
To me the diff looks like that code is only called when the first item is not empty? This is my point in the original comment (5 days ago).

Reply:
Right. This should check `result` instead of `dss`.

Reply:
... which brings us back to my earlier point: It feels like the "invalid dataset" solution is way to easy to mess up, even for you as the original author?